File: esw-integration-test/src/test/scala/esw/gateway/server/testdata/SampleComponentHandlers.scala
package esw.gateway.server.testdata

import akka.actor.typed.scaladsl.ActorContext
import csw.command.client.messages.TopLevelActorMessage
import csw.framework.models.CswContext
import csw.framework.scaladsl.ComponentHandlers
import csw.location.models.TrackingEvent
import csw.logging.api.scaladsl.Logger
import csw.params.commands.CommandResponse.{Accepted, Completed}
import csw.params.commands.{CommandResponse, ControlCommand}
import csw.params.core.generics.KeyType.StringKey
import csw.params.core.models.Prefix
import csw.params.core.states.{CurrentState, StateName}
import csw.params.events.{EventName, SystemEvent}
import csw.time.core.models.UTCTime

import scala.concurrent.Future

class SampleComponentHandlers(ctx: ActorContext[TopLevelActorMessage], cswCtx: CswContext)
    extends ComponentHandlers(ctx, cswCtx) {
  import cswCtx._

  val log: Logger = loggerFactory.getLogger(ctx)
  override def initialize(): Future[Unit] = {
    log.info("Initializing Component TLA")
    Thread.sleep(100)
    Future.unit
  }

  override def onLocationTrackingEvent(trackingEvent: TrackingEvent): Unit = ???

  override def validateCommand(controlCommand: ControlCommand): CommandResponse.ValidateCommandResponse =
    Accepted(controlCommand.runId)

  override def onSubmit(controlCommand: ControlCommand): CommandResponse.SubmitResponse =
    Completed(controlCommand.runId)

  override def onOneway(controlCommand: ControlCommand): Unit = {
    val currentState1 = CurrentState(Prefix("esw.a.b"), StateName("stateName1"))
    val currentState2 = CurrentState(Prefix("esw.a.b"), StateName("stateName2"))
    currentStatePublisher.publish(currentState1)
    currentStatePublisher.publish(currentState2)
    log.info("Invoking Oneway Handler TLA")
  }

  override def onShutdown(): Future[Unit] = {
    log.info("Shutting down Component TLA")
    Future.unit
  }

  override def onGoOffline(): Unit = {
    val event = new SystemEvent(Prefix("tcs.filter.wheel"), EventName("offline"))
    eventService.defaultPublisher.publish(event)
  }

  override def onGoOnline(): Unit = {
    val event = new SystemEvent(Prefix("tcs.filter.wheel"), EventName("online"))
    eventService.defaultPublisher.publish(event)
  }

  override def onDiagnosticMode(startTime: UTCTime, hint: String): Unit = {
    val diagnosticModeParam = StringKey.make("mode").set("diagnostic")
    val event               = SystemEvent(Prefix("tcs.filter.wheel"), EventName("diagnostic-data")).add(diagnosticModeParam)
    eventService.defaultPublisher.publish(event)
  }

  override def onOperationsMode(): Unit = {
    val operationsModeParam = StringKey.make("mode").set("operations")
    val event               = SystemEvent(Prefix("tcs.filter.wheel"), EventName("diagnostic-data")).add(operationsModeParam)
    eventService.defaultPublisher.publish(event)
  }
}

File: esw-kt/integration-test/src/test/java/esw/ocs/testdata/SampleComponentHandlers.java
package esw.ocs.testdata;

import akka.actor.typed.javadsl.ActorContext;
import csw.command.client.messages.TopLevelActorMessage;
import csw.framework.javadsl.JComponentHandlers;
import csw.framework.models.JCswContext;
import csw.location.models.TrackingEvent;
import csw.logging.api.javadsl.ILogger;
import csw.params.commands.CommandResponse;
import csw.params.commands.CommandResponse.Accepted;
import csw.params.commands.CommandResponse.Completed;
import csw.params.commands.ControlCommand;
import csw.params.core.generics.Parameter;
import csw.params.core.models.Prefix;
import csw.params.events.EventName;
import csw.params.events.SystemEvent;
import csw.params.javadsl.JKeyType;
import csw.time.core.models.UTCTime;

import java.util.concurrent.CompletableFuture;

public class SampleComponentHandlers extends JComponentHandlers {
    ActorContext<TopLevelActorMessage> ctx;
    JCswContext cswCtx;
    ILogger log;
    public SampleComponentHandlers(ActorContext<TopLevelActorMessage> ctx, JCswContext cswCtx) {
        super(ctx, cswCtx);
        this.cswCtx = cswCtx;
        this.ctx = ctx;
        this.log = cswCtx.loggerFactory().getLogger(ctx, getClass());
    }

    @Override
    public CompletableFuture<Void> jInitialize() {
        log.info("Initializing Component TLA");
        try {
            Thread.sleep(100);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletableFuture<Void> jOnShutdown() {
        log.info("Shutting down Component TLA");
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public void onLocationTrackingEvent(TrackingEvent trackingEvent) {

    }

    @Override
    public CommandResponse.ValidateCommandResponse validateCommand(ControlCommand controlCommand) {
        return new Accepted(controlCommand.runId());
    }

    @Override
    public CommandResponse.SubmitResponse onSubmit(ControlCommand controlCommand) {
        return new Completed(controlCommand.runId());
    }

    @Override
    public void onOneway(ControlCommand controlCommand) {

    }

    @Override
    public void onDiagnosticMode(UTCTime startTime, String hint) {
        Parameter<String> diagnosticModeParam = JKeyType.StringKey().make("mode").set("diagnostic");
        SystemEvent event = new SystemEvent(Prefix.apply("tcs.filter.wheel"), EventName.apply("diagnostic-data")).add(diagnosticModeParam);
        cswCtx.eventService().defaultPublisher().publish(event);
    }

    @Override
    public void onOperationsMode() {
        Parameter<String> diagnosticModeParam = JKeyType.StringKey().make("mode").set("operations");
        SystemEvent event = new SystemEvent(Prefix.apply("tcs.filter.wheel"), EventName.apply("diagnostic-data")).add(diagnosticModeParam);
        cswCtx.eventService().defaultPublisher().publish(event);
    }

    @Override
    public void onGoOffline() {
        SystemEvent event = new SystemEvent(Prefix.apply("tcs.filter.wheel"), EventName.apply("offline"));
        cswCtx.eventService().defaultPublisher().publish(event);
    }

    @Override
    public void onGoOnline() {
        SystemEvent event = new SystemEvent(Prefix.apply("tcs.filter.wheel"), EventName.apply("online"));
        cswCtx.eventService().defaultPublisher().publish(event);
    }
}
